### DISCUSSION TITLE
Wire disconnects on screwdriver unequip

### DISCUSSION THREAD

**Discussion by Androidonator** [FakeFishGames/Barotrauma#13805] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

when you want to wire circuit box, but the wire disconnects after you unequip screwdriver. returning to lobby didnt fix it happened in multiplayer

It happens randomly not every server you join and have wire rights.
Even owner tried didnt work for him :(

### Reproduction steps

1. join multiplayer
2. think about some cool circuit you want to make (not mandatory)
3. connect circuit box input to anything
4. unequip screwdriver
5. wire disconnects and drops on the ground

### Bug prevalence

Happens every now and then

### Single player or multiplayer?

Multiplayer hosted from the in-game menu (= using a listen server)

### Version

v1.3.0.4

### Which operating system did you encounter this bug on?

Windows

**Comment by RodionGromo** [comment]
Happened too, tried to connect some wires on Orca2, they were just falling on screwdriver unequip

**Comment by Saveforth** [comment]
I've had the same issue. Windows 11, server, modded. Wires worked for a couple missions and then suddenly the wires won't stay in place if I unequip the screwdriver or equip a wire.

**Comment by Shmepp** [comment]
This is still an issue. Has anyone found any fixes or work arounds?

**Reply by evonichr4** [reply]
> This is still an issue. Has anyone found any fixes or work arounds?

I had sworn I took no wires from a previous structure or ship, but after getting rid of all my wires I found that buying new ones worked fine. So try buying new wires (and don't stack them with any wires you own) and see if that works.

**Comment by ADAMASTER89** [comment]
I've tried to reproduce the issue with the provided steps and some additional scenarios around them but could not get the issue to occur.
I'd like to ask you to record and share a video whenever you encounter it again.

**Reply by Androidonator** [reply]
This has been so long ago. I will loook into it but I don't even know how to get any sort of dev mode running.

**Comment by Ivanzyfer** [comment]
Reproduced in single player with no mods, 1.15.19.0 + Home & Harbor. The wires that behave like this are ordinary items carrying a saved `DropOnConnect = false`, and every one of them comes from a circuit box.

**What happens**

A wire is taken out of your hand by exactly one place, `Wire.TryConnect` (`Wire.cs:920`):

```cs
if (DropOnConnect && connections[0] != null && connections[1] != null)
{
    foreach (ItemComponent component in item.Components)
    {
        if (component != this)
        {
            component.Drop(null);
        }
    }
    ...
}
```

`Pickable.Drop` (`Pickable.cs:291`) is what removes the item from the inventory, and the wire's own `Wire.Drop` is skipped, so the connections survive. With `DropOnConnect = false` none of this runs: the wire is connected at both ends but stays in your hand and is never spent from the stack. From the player's side this looks like the second end refusing to connect, exactly as [15941](https://github.com/FakeFishGames/Barotrauma/discussions/15941) puts it.

After that `Wire.Unequip` (`Wire.cs:1017`) clears both ends unconditionally, while `Equip` and `Drop` are guarded by `shouldClearConnections`. That is the "wire disconnects when I unequip the screwdriver or equip a new wire" from the reports.

**Where those wires come from**

`DropOnConnect` is saved with the item (`Wire.cs:216`, `IsPropertySaveable.Yes`) and is set to false for the wires inside a circuit box (`CircuitBoxWire.cs:123`).

I parsed all 234 `.sub` files of the vanilla game and the DLC: `DropOnConnect="False"` appears 254 times, and every occurrence is inside a circuit box, none outside. Almost all of them are `bluewire`, which is why blue wires get noticed and green ones do not. Counts: BeaconStation_Hissihiss 28, its shuttle 26, NewIapetus 15, player subs 4 to 12 (Dugong 10), Dugong_NPC 3, Orca_NPC 5.

They leave the box in two vanilla ways:

- deconstructing the box: contained items are moved to the output slots (`Deconstructor.cs:706-720`), and the wire container on `circuitbox` is not marked `RemoveContainedItemsOnDeconstruct`;
- destroying the box: `ItemContainer.RemoveComponentSpecific` (`ItemContainer.cs:1881`) drops everything inside onto the floor.

From then on the wire is indistinguishable from a normal one and stacks with it (`maxstacksize` 32, 8 in hands). That is why the behaviour looks random and why a whole stack seems to go bad, which is what [15941](https://github.com/FakeFishGames/Barotrauma/discussions/15941) describes.

**Steps, single player, no mods**

1. A campaign with a sub that has a circuit box. Dugong has one, with ten blue wires inside.
2. Detach the box with a wrench and deconstruct it at an outpost.
3. Take the blue wires from the output slots and keep them apart from your own.
4. Wire two devices together with one of them.
5. The wire is not spent: it stays in your hand and the stack count does not change. The connection does show up in the panel at first.
6. Leave the device and open it again, or put anything else in that hand: the wire is no longer in the connection.
7. Control: repeat with a blue wire from the sub's cabinet, it leaves your hand as usual.

No video is needed for this: the difference shows on two visually identical blue wires from the same sub.

This looks like the same cause behind [14076](https://github.com/FakeFishGames/Barotrauma/discussions/14076) and [15941](https://github.com/FakeFishGames/Barotrauma/discussions/15941): the same sign there, the wire stays in hand and a whole stack goes bad. If so, both can be closed together with this one.
